docs/examples/expressions.ipynb

```diff
@@ -0,0 +1,170 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Defining the iterator\n",
+    "\n",
+    "We will use custom iterator producing small tensors filled with series of numbers, so we can easily inspect the results."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class ExternalInputIterator(object):\n",
+    "    def __init__(self, batch_size, left_type, right_type):\n",
+    "        self.batch_size = batch_size\n",
+    "        self.left_type = left_type\n",
+    "        self.right_type = right_type\n",
+    "\n",
+    "    def __iter__(self):\n",
+    "        self.i = 0\n",
+    "        self.n = 128\n",
+    "        return self\n",
+    "\n",
+    "    def __next__(self):\n",
+    "        left = []\n",
+    "        right = []\n",
+    "        for sample in range(self.batch_size):\n",
+    "            left.append(np.array([sample + self.i], dtype = self.left_type))\n",
+    "            right.append(np.array([self.batch_size + self.i + sample], dtype = self.right_type))\n",
+    "        self.i = (self.i + 1) % self.n\n",
+    "        return (left, right)\n",
+    "    \n",
+    "    next = __next__"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Defining the pipeline\n",
+    "\n",
+    "The next step is to define the Pipeline.\n",
+    "\n",
+    "We override `Pipeline.iter_setup`, a method called by the pipeline before every `Pipeline.run`, to call the iterator\n",
+    "and feed the result to `ExternalSource()` operator, referenced by `self.jpeg`, by using `feed_input`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "iterator_u8_u8 = iter(ExternalInputIterator(batch_size, np.uint8, np.uint8))\n",
+    "iterator_u8_i32 = iter(ExternalInputIterator(batch_size, np.uint8, np.int32))\n",
+    "iterator_i16_u8 = iter(ExternalInputIterator(batch_size, np.int16, np.uint8))\n",
+    "iterator_i32_f32 = iter(ExternalInputIterator(batch_size, np.int32, np.float32))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    " class ExternalSourcePipeline(Pipeline):                   \n",
+    "    def __init__(self, iterator, batch_size, num_threads, device_id):\n",
+    "        super(ExternalSourcePipeline, self).__init__(batch_size, num_threads, device_id, seed=12)\n",
+    "        self.left_source = ops.ExternalSource()\n",
+    "        self.right_source = ops.ExternalSource()\n",
+    "        self.iterator = iterator\n",
+    "\n",
+    "    def define_graph(self):                                                                \n",
+    "        self.left = self.left_source()\n",
+    "        self.right = self.right_source()\n",
+    "        return self.left, self.right, self.left + self.right * self.left\n",
```

Comment:
I would add a comment that now you can operate on tensors in `define_graph` and DALI under the hood would insert right operators.

Reply:
Done